server/ws/websockets.go

```diff
@@ -154,20 +109,123 @@ func (ws *Server) handleWebSocketOnChange(w http.ResponseWriter, r *http.Request
+			ws.authenticateListener(&wsSession, command.Token)
+
 		case "ADD":
 			log.Printf(`Command: Add blockID: %v, client: %s`, command.BlockIDs, client.RemoteAddr())
-			ws.AddListener(client, command.BlockIDs)
+			ws.addListener(&wsSession, command.BlockIDs)
 
 		case "REMOVE":
 			log.Printf(`Command: Remove blockID: %v, client: %s`, command.BlockIDs, client.RemoteAddr())
-			ws.RemoveListenerFromBlocks(client, command.BlockIDs)
+			ws.removeListenerFromBlocks(&wsSession, command.BlockIDs)
 
 		default:
 			log.Printf(`ERROR webSocket command, invalid action: %v`, command.Action)
 		}
 	}
 }
 
+func (ws *Server) isValidSessionToken(token string) bool {
+	if ws.singleUser {
+		return true
+	}
+
+	session, err := ws.auth.GetSession(token)
+	if session == nil || err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (ws *Server) authenticateListener(wsSession *websocketSession, token string) {
+	isValidSession := ws.isValidSessionToken(token)
+	if !isValidSession {
+		wsSession.client.Close()
+		return
+	}
+
+	// Authenticated
+	wsSession.isAuthenticated = true
+	log.Printf("authenticateListener: Authenticated")
+}
+
+// AddListener adds a listener for a block's change.
+func (ws *Server) addListener(wsSession *websocketSession, blockIDs []string) {
+	if !wsSession.isAuthenticated {
+		log.Printf("addListener: NOT AUTHENTICATED")
+		return
+	}
+
+	ws.mu.Lock()
+	for _, blockID := range blockIDs {
+		if ws.listeners[blockID] == nil {
+			ws.listeners[blockID] = []*websocket.Conn{}
+		}
+
+		ws.listeners[blockID] = append(ws.listeners[blockID], wsSession.client)
+	}
+	ws.mu.Unlock()
+}
+
+// removeListener removes a webSocket listener from all blocks.
+func (ws *Server) removeListener(client *websocket.Conn) {
+	ws.mu.Lock()
+	for key, clients := range ws.listeners {
+		listeners := []*websocket.Conn{}
+
+		for _, existingClient := range clients {
+			if client != existingClient {
+				listeners = append(listeners, existingClient)
+			}
+		}
+
+		ws.listeners[key] = listeners
+	}
+	ws.mu.Unlock()
+}
+
+// removeListenerFromBlocks removes a webSocket listener from a set of block.
+func (ws *Server) removeListenerFromBlocks(wsSession *websocketSession, blockIDs []string) {
+	if !wsSession.isAuthenticated {
+		log.Printf("removeListenerFromBlocks: NOT AUTHENTICATED")
+		return
+	}
+
+	ws.mu.Lock()
+
+	for _, blockID := range blockIDs {
+		listeners := ws.listeners[blockID]
+		if listeners == nil {
+			return
+		}
+
+		// Remove the first instance of this client that's listening to this block
+		// Note: A client can listen multiple times to the same block
+		for index, listener := range listeners {
+			if wsSession.client == listener {
+				newListeners := append(listeners[:index], listeners[index+1:]...)
+				ws.listeners[blockID] = newListeners
+
+				break
+			}
+		}
+	}
+
+	ws.mu.Unlock()
+}
+
+// GetListeners returns the listeners to a blockID's changes.
+func (ws *Server) GetListeners(blockID string) []*websocket.Conn {
```

Comment:
```suggestion
// getListeners returns the listeners to a blockID's changes.
func (ws *Server) getListeners(blockID string) []*websocket.Conn {
```
and of course is needed to change the call to `GetListeners` called later.

Reply:
Thanks. Cleaned up the function visibility.